src/scope.rs

```diff
@@ -225,6 +225,153 @@ impl BasePattern {
 
         true
     }
+
+    /// Get an iterator over all identifiers inside the pattern.
+    pub fn identifiers(&self) -> impl Iterator<Item = &Identifier> {
+        self.pre_order_iter().flat_map(BasePattern::as_identifier)
+    }
+
+    /// Check if all `identifiers` are contained inside the pattern.
+    pub fn contains_all<'a, I>(&self, mut identifiers: I) -> bool
+    where
+        I: Iterator<Item = &'a Identifier>,
+    {
+        identifiers.all(|id| self.contains(id))
+    }
```

Comment:
In 8477e265153d4864f228caa4c819e77f91e678bb:

This implementation of `contains_all` also has bad asymptotics. It shouldn't call `contains` repeatedly. Instead you should construct a hashset of all the identifiers in `self` and then check that every item yielded from the iterator lands in the hashset.

Reply:
`contains` iterates over the tree in pre-order short-citcuits on the first match. This is exactly what `DirectedTree` does.

The problem is `contains_all` which repeatedly calls `contains`, making it O(n^2). Let's address the performance in a follow-up PR. Thanks for your suggestion to use a hashset.

Reply:
Oh, I see that `get` also calls `contains`. This is indeed slower than `DirectedTree`. Again, thanks for your suggested fix. We might need different `contains` methods for `contains_all` and for `get`.